

```
# This is formatted as code
```

# DMA Fall 20

**Note** : This entire lab will be manually evaluated.

Name : 'Yvonne Zhou'

Collaborator : ''

# Lab 4: Neural Networks #


In [ ]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.feature_extraction import DictVectorizer

from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, ParameterGrid

import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [ ]:
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_training.csv
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_test.csv

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')
df_train.head()

URL transformed to HTTPS due to an HSTS policy
--2020-09-23 18:46:06--  https://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_training.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 105581 (103K) [text/csv]
Saving to: ‘lab_4_training.csv.12’

lab_4_training.csv. 100%[===================>] 103.11K  --.-KB/s    in 0.06s   

2020-09-23 18:46:06 (1.61 MB/s) - ‘lab_4_training.csv.12’ saved [105581/105581]

URL transformed to HTTPS due to an HSTS policy
--2020-09-23 18:46:06--  https://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_test.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2

Unnamed: 0 gender  age    year  ... exercisehours  musiccds  playgames  watchtv
0        1303   male   20  second  ...           5.0      50.0        1.0     15.0
1          36   male   20   third  ...           4.0      10.0        0.0      1.0
2         489   male   22  fourth  ...           2.0     150.0        1.0     10.0
3        1415   male   19  second  ...           5.0     100.0        0.0      7.0
4         616   male   22  fourth  ...           7.0      10.0        0.0      5.0

[5 rows x 15 columns]

***
### Question 1###
Calculate a baseline accuracy measure using the majority class, assuming a target variable of 'gender'. The majority class is the most common value of the target variable in a particular dataset. Accuracy is calculated as (true positives + true negatives) / (all negatives and positives)

**Question 1.a**  
Find the majority class in the training set. If you always predicted this class in the training set, what would your accuracy be?

In [ ]:
# YOUR CODE HERE
majority = df_train.groupby("gender").count().reset_index().sort_values(by = "Unnamed: 0", ascending = False).iloc[0, 0]
train_gender_with_prediction = df_train.copy()
train_gender_with_prediction["all_majority"] = majority
train_gender_with_prediction = train_gender_with_prediction[["gender", "all_majority"]]
train_all_majority_accuracy = train_gender_with_prediction[train_gender_with_prediction["gender"] == 
                                                           train_gender_with_prediction["all_majority"]].shape[0] / train_gender_with_prediction.shape[0]
print("Majority Class:", majority, "; Accuracy:", train_all_majority_accuracy)

Majority Class: female ; Accuracy: 0.5377358490566038


###ANSWER: The majority class is female, and the accuracy of always predict this class in the training set is 0.5377358490566038.

**Question 1.b**   
If you always predicted this same class (majority from the training set) in the test set, what would your accuracy be?

In [ ]:
# YOUR CODE HERE
test_gender_with_prediction = df_test.copy()
test_gender_with_prediction["all_majority"] = majority
test_gender_with_prediction = test_gender_with_prediction[["gender", "all_majority"]]
test_all_majority_accuracy = test_gender_with_prediction[test_gender_with_prediction["gender"] == 
                                                         test_gender_with_prediction["all_majority"]].shape[0] / test_gender_with_prediction.shape[0]
print("Accuracy:", test_all_majority_accuracy)

Accuracy: 0.5226130653266332


###ANSWER: The accuracy of always predict this majority class in the test set is 0.5226130653266332.

***
### Question 2 ###
Get started with Neural Networks.

   
Choose a NN implementation (eg: scikit-learn) and specify which you choose. Be sure the implementation allows you to modify the number of hidden layers and hidden nodes per layer.  

NOTE: When possible, specify the logsig (sigmoid/logistc) function as the transfer function (another word for activation function) and use Levenberg-Marquardt backpropagation (lbfgs). It is possible to specify logistic in Sklearn MLPclassifier (Neural net).  

**Question 2.a**   
Train a neural network with a single 10 node hidden layer. Only use the Height feature of the dataset to predict the Gender. You will have to change Gender to a 0 and 1 class. After training, use your trained model to predict the class using the height feature from the training set. What was the accuracy of this prediction?

In [ ]:
# YOUR CODE HERE
df_train_modified = df_train.copy()
df_train_modified["gender"] = df_train_modified["gender"].replace("female", 1)
df_train_modified["gender"] = df_train_modified["gender"].replace("male", 0)

clf_2a = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', random_state = 42)
clf_2a.fit(df_train_modified[["height"]], df_train_modified["gender"])
clf_2a.score(df_train_modified[["height"]], df_train_modified["gender"])

0.6440251572327044

###ANSWER: The accuracy of this prediction is 0.6440251572327044.

**Question 2.b**  
Take the trained model from question 2.a and use it to predict the test set. This can be accomplished by taking the trained model and giving it the Height feature values from the test set. What is the accuracy of this model on the test set?

In [ ]:
# YOUR CODE HERE
df_test_modified = df_test.copy()
df_test_modified["gender"] = df_test_modified["gender"].replace("female", 1)
df_test_modified["gender"] = df_test_modified["gender"].replace("male", 0)
clf_2a.score(df_test_modified[["height"]], df_test_modified["gender"])

0.6206030150753769

###ANSWER: The accuracy of this model on the test set is 0.6206030150753769.

**Question 2.c**   
Neural Networks tend to prefer smaller, normalized feature values. Try taking the log of the height feature in both training and testing sets or use a Standard Scalar operation in SKlearn to centre and normalize the data between 0-1 for continuous values. Repeat question 2.a and 2.b with the log version or the normalized and centered version of this feature

In [ ]:
# YOUR CODE HERE
log_df_train = df_train.copy()
log_df_train["height"] = np.log(log_df_train["height"])
log_df_test = df_test.copy()
log_df_test["height"] = np.log(log_df_test["height"])

log_df_train["gender"] = log_df_train["gender"].replace("female", 1)
log_df_train["gender"] = log_df_train["gender"].replace("male", 0)

clf_2c = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', random_state = 42)
clf_2c.fit(log_df_train[["height"]], log_df_train["gender"])

log_df_test["gender"] = log_df_test["gender"].replace("female", 1)
log_df_test["gender"] = log_df_test["gender"].replace("male", 0)

print("training accuracy:", clf_2c.score(log_df_train[["height"]], log_df_train["gender"]), 
      "; testing accuracy:", clf_2c.score(log_df_test[["height"]], log_df_test["gender"]))

training accuracy: 0.5389937106918239 ; testing accuracy: 0.5226130653266332


###ANSWER: The training accuracy using this logged height feature is 0.5389937106918239, and the testing accuracy using this logged is 0.5226130653266332.

***

### Question 3###
The rest of features in this dataset barring a few are categorical. No ML method accepts categorical features, so transform year, eyecolor, exercise into a set of binary features, one feature per unique original feature value, and mark the binary feature as ‘1’ if the feature value matches the original value and ‘0’ otherwise. Using only these binary variable transformed features, train and predict the class of the test set. What was your accuracy using Neural Network with a single 10 node hidden layer? During training, use a maximum number of iterations of 50.

In [ ]:
# YOUR CODE HERE
selected_3_train = df_train[["year", "eyecolor", "exercise"]].replace({'first"': "first"})
selected_3_train_hot = pd.concat([pd.get_dummies(selected_3_train[["year"]], prefix = "year"), 
                                  pd.get_dummies(selected_3_train[["eyecolor"]], prefix = "eyecolor"),
                                  pd.get_dummies(selected_3_train[["exercise"]], prefix = "exercise")], axis = 1)

selected_3_test = df_test[["year", "eyecolor", "exercise"]].replace({'first"': "first"})
selected_3_test_hot = pd.concat([pd.get_dummies(selected_3_test[["year"]], prefix = "year"), 
                                 pd.get_dummies(selected_3_test[["eyecolor"]], prefix = "eyecolor"),
                                 pd.get_dummies(selected_3_test[["exercise"]], prefix = "exercise")], axis = 1)
"""
enc.fit(selected_3)
one_hot_labeled = pd.DataFrame(enc.transform(selected_3).toarray())
one_hot_labeled = one_hot_labeled.rename({0: 0, 1: 1, 2: "year_fourth", 3: 3, 4: "year_second", 5:, "year_third", 
                                          6: "eyecolor_blue", 7: "eyecolor_brown", 8: "eyecolor_green", 9: "eyecolor_hazel", 10: "eyecolor_other",
                                          11: "exercise_no", 12: "exercise_yes"})
"""
clf_3 = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', max_iter = 50, random_state = 42)
clf_3.fit(selected_3_train_hot, df_train_modified["gender"])

print("training accuracy:", clf_3.score(selected_3_train_hot, df_train_modified["gender"]), 
      "; testing accuracy:", clf_3.score(selected_3_test_hot, df_test_modified["gender"]))

training accuracy: 0.5679245283018868 ; testing accuracy: 0.6030150753768844


###ANSWER: The training accuracy using Neural Network with the one hot coded features is 0.5679245283018868, and the testing accuracy is 0.6030150753768844.

***
### Question 4###
Using a NN, report the accuracy on  the test set of a model that trained only on the height and the eye color features of instances in the training set.

**Question 4.a**  
What is the accuracy on the test set using the original height values (no pre-processing) and eye color as a one-hot?

In [ ]:
# YOUR CODE HERE
columns_4a = ["eyecolor_blue", "eyecolor_brown", "eyecolor_green", "eyecolor_hazel", "eyecolor_other"]

train_height_and_eye = pd.concat([selected_3_train_hot[columns_4a], df_train_modified["height"]], axis = 1)
test_height_and_eye = pd.concat([selected_3_test_hot[columns_4a], df_test_modified["height"]], axis = 1)

clf_4a = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', max_iter = 50, random_state = 42)
clf_4a.fit(train_height_and_eye, df_train_modified["gender"])

print("training accuracy:", clf_4a.score(train_height_and_eye, df_train_modified["gender"]), 
      "; testing accuracy:", clf_4a.score(test_height_and_eye, df_test_modified["gender"]))

training accuracy: 0.5377358490566038 ; testing accuracy: 0.5226130653266332


###ANSWER: The training accuracy using the original height values and eye color as a one-hot is 0.5377358490566038, and the testing accuracy is 0.5226130653266332.

**Question 4.b**  
What is the accuracy on the test set using the log of height values (applied to both training and testing sets) and eye color as a one-hot?

In [ ]:
# YOUR CODE HERE
train_log_height_and_eye = pd.concat([selected_3_train_hot[columns_4a], log_df_train["height"]], axis = 1)
test_log_height_and_eye = pd.concat([selected_3_test_hot[columns_4a], log_df_test["height"]], axis = 1)

clf_4b = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', max_iter = 50, random_state = 42)
clf_4b.fit(train_log_height_and_eye, log_df_train["gender"])

print("training accuracy:", clf_4b.score(train_log_height_and_eye, log_df_train["gender"]), 
      "; testing accuracy:", clf_4b.score(test_log_height_and_eye, log_df_test["gender"]))

training accuracy: 0.8232704402515724 ; testing accuracy: 0.8291457286432161


###ANSWER: The training accuracy using the logged height values and eye color as a one-hot is 0.8232704402515724, and the testing accuracy is 0.8291457286432161.

**Question 4.c**  
What is the accuracy on the test set using the Z-score of height values and eye color as a one-hot? 

Z-score is a normalization function. It is the value of a feature minus the average value for that feature (in the training set), divided by the standard deviation of that feature (in the training set). Remember that, whenever applying a function to a feature in the training set, it also has to be applied to that same feature in the test set.

In [ ]:
# YOUR CODE HERE
from scipy.stats import zscore

train_z_height_and_eye = pd.concat([selected_3_train_hot[columns_4a], pd.Series(zscore(df_train_modified["height"]))], axis = 1)
test_z_height_and_eye = pd.concat([selected_3_test_hot[columns_4a], pd.Series(zscore(df_test_modified["height"]))], axis = 1)

clf_4c = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', max_iter = 50, random_state = 42)
clf_4c.fit(train_z_height_and_eye, df_train_modified["gender"])

print("training accuracy:", clf_4c.score(train_z_height_and_eye, df_train_modified["gender"]), 
      "; testing accuracy:", clf_4c.score(test_z_height_and_eye, df_test_modified["gender"]))

training accuracy: 0.8490566037735849 ; testing accuracy: 0.8567839195979899


###ANSWER: The training accuracy using the z-score height values and eye color as a one-hot is 0.8490566037735849, and the testing accuracy is 0.8567839195979899.

***
### Question 5 ###
Repeat question 4 for exercise hours + eye color

In [ ]:
# YOUR CODE HERE
# no pre-processing
columns_5 = ["eyecolor_blue", "eyecolor_brown", "eyecolor_green", "eyecolor_hazel", "eyecolor_other"]

train_exercisehours_and_eye = pd.concat([selected_3_train_hot[columns_5], df_train_modified["exercisehours"]], axis = 1)
test_exercisehours_and_eye = pd.concat([selected_3_test_hot[columns_5], df_test_modified["exercisehours"]], axis = 1)

clf_5a = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', max_iter = 50, random_state = 42)
clf_5a.fit(train_exercisehours_and_eye, df_train_modified["gender"])

print("training accuracy for no pre-processing:", clf_5a.score(train_exercisehours_and_eye, df_train_modified["gender"]), 
      "; testing accuracy for no pre-processing:", clf_5a.score(test_exercisehours_and_eye, df_test_modified["gender"]))

# logged exercisehours
train_log_exercisehours_and_eye = pd.concat([selected_3_train_hot[columns_5], np.log(df_train["exercisehours"].replace({0: 0.000000001}))], axis = 1)
test_log_exercisehours_and_eye = pd.concat([selected_3_test_hot[columns_5], np.log(df_test["exercisehours"].replace({0: 0.000000001}))], axis = 1)

clf_5b = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', max_iter = 50, random_state = 42)
clf_5b.fit(train_log_exercisehours_and_eye, df_train_modified["gender"])

print("training accuracy for logged exercisehours:", clf_5b.score(train_log_exercisehours_and_eye, log_df_train["gender"]), 
      "; testing accuracy for logged exercisehours:", clf_5b.score(test_log_exercisehours_and_eye, log_df_test["gender"]))

# z-scored exercisehours
train_z_exercisehours_and_eye = pd.concat([selected_3_train_hot[columns_5], pd.Series(zscore(df_train_modified["exercisehours"]))], axis = 1)
test_z_exercisehours_and_eye = pd.concat([selected_3_test_hot[columns_5], pd.Series(zscore(df_test_modified["exercisehours"]))], axis = 1)

clf_5c = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', max_iter = 50, random_state = 42)
clf_5c.fit(train_z_exercisehours_and_eye, df_train_modified["gender"])

print("training accuracy for z-scored exercisehours:", clf_5c.score(train_z_exercisehours_and_eye, df_train_modified["gender"]), 
      "; testing accuracy for z-scored exercisehours:", clf_5c.score(test_z_exercisehours_and_eye, df_test_modified["gender"]))

training accuracy for no pre-processing: 0.5867924528301887 ; testing accuracy for no pre-processing: 0.5753768844221105
training accuracy for logged exercisehours: 0.5823899371069182 ; testing accuracy for logged exercisehours: 0.5728643216080402
training accuracy for z-scored exercisehours: 0.5874213836477987 ; testing accuracy for z-scored exercisehours: 0.5653266331658291


###ANSWER: 
###The training accuracy using the no pre-processing exercisehours values and eye color as a one-hot is 0.5867924528301887, and the testing accuracy is 0.5753768844221105. 
###The training accuracy using the logged exercisehours values and eye color as a one-hot is 0.5823899371069182, and the testing accuracy is 0.5728643216080402. 
###The training accuracy using the z-scored exercisehours values and eye color as a one-hot is 0.5874213836477987, and the testing accuracy is 0.5653266331658291.


***
### Question 6###
Combine the features from question 3, 4, and 5 (year, eyecolor, exercise, height, exercise hours). For numeric features use the best normalization method from questions 4 and 5.

**Question 6.a**  
What was the NN accuracy on the test set using the single 10 node hidden layer?

In [ ]:
# YOUR CODE HERE
train_z_all = pd.concat([selected_3_train_hot, 
                         pd.Series(zscore(df_train_modified["height"])), 
                         pd.Series(zscore(df_train_modified["exercisehours"]))], axis = 1)
test_z_all = pd.concat([selected_3_test_hot, 
                        pd.Series(zscore(df_test_modified["height"])), 
                        pd.Series(zscore(df_test_modified["exercisehours"]))], axis = 1)

clf_6 = MLPClassifier(hidden_layer_sizes = (10, ), activation = "logistic", solver = 'lbfgs', max_iter = 50, random_state = 42)
clf_6.fit(train_z_all, df_train_modified["gender"])

print("testing accuracy:", clf_6.score(test_z_all, df_test_modified["gender"]))

testing accuracy: 0.8668341708542714


###ANSWER: The NN accuracy on the test set using the single 10 node hidden layer is 0.8668341708542714.

***
### Question 7- Bonus (10%)###
Can you improve your test set prediction accuracy by 5% or more?  

See how close to that milestone of improvement you can get by modifying the tuning parameters of  Neural Networks (the number of hidden layers, number of hidden nodes in each layer, the learning rate aka mu). A great guide to tuning parameters is explained in this guide: http://www.csie.ntu.edu.tw/~cjlin/papers/guide/guide.pdf. 

While the guide is specific to SVM and in particular the C and gamma parameters of the RBF kernel, the method applies to generally to any ML technique with tuning parameters.

Please also write a paragraph in a markdown cell below with an explanation of your approach and evaluation metrics.


In [ ]:
# YOUR CODE HERE
train_z_all = pd.concat([selected_3_train_hot, 
                         pd.Series(zscore(df_train_modified["height"])), 
                         pd.Series(zscore(df_train_modified["exercisehours"]))], axis = 1)
test_z_all = pd.concat([selected_3_test_hot, 
                        pd.Series(zscore(df_test_modified["height"])), 
                        pd.Series(zscore(df_test_modified["exercisehours"]))], axis = 1)

clf_7 = MLPClassifier(hidden_layer_sizes = (100, ), activation = "logistic", solver = 'lbfgs', max_iter = 1000, random_state = 42)
clf_7.fit(train_z_all, df_train_modified["gender"])

print("testing accuracy:", clf_7.score(test_z_all, df_test_modified["gender"]))

testing accuracy: 0.9246231155778895


###ANSWER: I increased the number of neurons in hidden layers and also the maximum number of iteration, and the accuracy improved about 6%. This is because as we are increasing the number of neurons and the number of iteration, we are increasing the possible complexity of our model. The model before I increased these two parameters are underfitting, so increasing the model's complexity would make the testing accuracy become higher.